File: business/src/main/java/com/pij/lottieshow/interactor/DirectorySource.java
package com.pij.lottieshow.interactor;

import com.pij.lottieshow.model.LottieFile;

import java.io.File;

import rx.Observable;

import static java.util.Collections.emptyList;
import static rx.Observable.just;

/**
 * <p>Created on 10/04/2017.</p>
 * @author Pierrejean
 */
@SuppressWarnings("WeakerAccess")
public class DirectorySource implements LottieSource {

    private final Observable<File> storageRoot;

    public DirectorySource(Observable<File> storageRoot) {
        this.storageRoot = storageRoot;
    }

    /**
     * Always provide an initial sequence, an empty one at worst.
     */
    @Override
    public Observable<Iterable<LottieFile>> lottieFiles() {
        Observable<File[]> files = storageRoot.map(root -> root.isDirectory() ? root.listFiles() : new File[0]);
        Observable<Iterable<LottieFile>> lotties = files.flatMap(list -> Observable.from(list).map(File::toURI)
                                                                                   .map(LottieFile::create)
                                                                                   .toList()
                                                                                   .doOnError
                                                                                           (Throwable::printStackTrace)
                                                                                   .onErrorReturn(e -> emptyList()));
        return lotties.startWith(just(emptyList())).distinctUntilChanged();
    }

}

File: business/src/main/java/com/pij/lottieshow/interactor/DirectorySource.kt
package com.pij.lottieshow.interactor

import com.pij.lottieshow.model.LottieFile
import com.pij.lottieshow.model.toLottie
import rx.Observable
import rx.Observable.just
import rx.Single
import rx.lang.kotlin.toSingle
import java.io.File
import java.net.URI

/**
 * Created on 10/04/2017.
 * @author Pierrejean
 */
class DirectorySource(private val storageRoot: Observable<File>) : LottieSource {

    /**
     * Always provide an initial sequence, an empty one at worst.
     */
    override fun lottieFiles(): Observable<Iterable<LottieFile>> {
        val files = storageRoot.map { root -> if (root.isDirectory) root.listFiles() else arrayOf() }
        val lotties = files.flatMapSingle(this::lottieFiles)
        return lotties.startWith(just(listOf())).distinctUntilChanged()
    }

    private fun lottieFiles(files: Array<File>): Single<Iterable<LottieFile>>? {
        return files.map(File::toURI)
                .map(URI::toLottie)
                .toSingle()
    }

}
